crates/ruma-common/src/push/condition/push_condition_serde.rs

```diff
@@ -0,0 +1,131 @@
+use serde::{de, Deserialize, Serialize, Serializer};
+use serde_json::value::RawValue as RawJsonValue;
+
+use crate::serde::from_raw_json_value;
+
+#[cfg(feature = "unstable-msc3931")]
+use super::RoomVersionFeature;
+use super::{PushCondition, RoomMemberCountIs};
+
+impl Serialize for PushCondition {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            PushCondition::_Custom(custom) => custom.serialize(serializer),
+            _ => PushConditionSerDeHelper::from(self.clone()).serialize(serializer),
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for PushCondition {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let json = Box::<RawJsonValue>::deserialize(deserializer)?;
+        let ExtractKind { kind } = from_raw_json_value(&json)?;
+
+        match kind.as_ref() {
+            "event_match"
+            | "contains_display_name"
+            | "room_member_count"
+            | "sender_notification_permission" => {
+                let helper: PushConditionSerDeHelper = from_raw_json_value(&json)?;
+                Ok(helper.into())
+            }
+            #[cfg(feature = "unstable-msc3931")]
+            "org.matrix.msc3931.room_version_supports" => {
+                let helper: PushConditionSerDeHelper = from_raw_json_value(&json)?;
+                Ok(helper.into())
+            }
+            _ => from_raw_json_value(&json).map(Self::_Custom),
+        }
+    }
+}
+
+#[derive(Deserialize)]
+struct ExtractKind {
+    kind: String,
+}
+
+#[derive(Serialize, Deserialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[serde(tag = "kind", rename_all = "snake_case")]
+enum PushConditionSerDeHelper {
```

Comment:
Do we have any existing tests for push condition (de)erialization?

Reply:
Yes, at the bottom of `condition.rs`